authority/provisioner/webhook.go

```diff
@@ -0,0 +1,204 @@
+	"go.step.sm/crypto/sshutil"
+	"golang.org/x/crypto/ssh"
+)
+
+type Webhook struct {
+	ID                   string `json:"id"`
+	Name                 string `json:"name"`
+	URL                  string `json:"url"`
+	Kind                 string `json:"kind"`
+	DisableTLSClientAuth bool   `json:"disableTLSClientAuth,omitempty"`
+	SigningSecret        string `json:"-"`
+	BearerToken          string `json:"-"`
+	BasicAuth            struct {
+		Username string
+		Password string
+	} `json:"-"`
+}
+
+type sshRequest struct {
+	Key        []byte
+	Type       string
+	KeyID      string
+	Principals []string
+}
+
+// ssh.Certificate but the Key and SignatureKey are marshaled
+type sshCert struct {
+	Nonce    []byte
+	Key      []byte
+	Serial   uint64
+	CertType uint32
+	//nolint:revive // field name matches ssh.Certificate
+	KeyId           string
+	ValidPrincipals []string
+	ValidAfter      uint64
+	ValidBefore     uint64
+	Permissions     ssh.Permissions
+	Reserved        []byte
+	SignatureKey    []byte
+	Signature       *ssh.Signature
+}
+
+type webhookRequestBody struct {
+	Timestamp string `json:"timestamp"`
+	//nolint:revive // acronyms can use all caps
+	X509_CSR []byte `json:"csr,omitempty"`
+	//nolint:revive // acronyms can use all caps
+	SSH_CR         *sshRequest       `json:"ssh_cr,omitempty"`
+	Certificate    *x509.Certificate `json:"certificate,omitempty"`
+	SSHCertificate *sshCert          `json:"ssh_certificate,omitempty"`
+}
+
+type WebhookResponseBody struct {
+	Data  map[string]interface{} `json:"data"`
+	Allow bool                   `json:"allow"`
+}
+
+func (w *Webhook) Do(ctx context.Context, client *http.Client, certReq interface{}, data map[string]interface{}) (*WebhookResponseBody, error) {
+	tmpl, err := template.New("url").Funcs(templates.StepFuncMap()).Parse(w.URL)
+	if err != nil {
+		return nil, err
+	}
+	buf := &bytes.Buffer{}
+	if err := tmpl.Execute(buf, data); err != nil {
+		return nil, err
+	}
+	url := buf.String()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	retries := 1
+retry:
+	reqBody := &webhookRequestBody{
+		Timestamp: time.Now().Format(time.RFC3339Nano),
+	}
+	switch r := certReq.(type) {
+	case *x509.CertificateRequest:
+		if r != nil {
+			reqBody.X509_CSR = r.Raw
+		}
+	case sshutil.CertificateRequest:
+		reqBody.SSH_CR = &sshRequest{
+			Type:       r.Type,
+			KeyID:      r.KeyID,
+			Principals: r.Principals,
+		}
+		if r.Key != nil {
+			reqBody.SSH_CR.Key = r.Key.Marshal()
+		}
+	case *x509.Certificate:
+		reqBody.Certificate = r
+	case *ssh.Certificate:
+		reqBody.SSHCertificate = &sshCert{
+			Nonce:           r.Nonce,
+			Key:             r.Key.Marshal(),
+			Serial:          r.Serial,
+			CertType:        r.CertType,
+			KeyId:           r.KeyId,
+			ValidPrincipals: r.ValidPrincipals,
+			ValidAfter:      r.ValidAfter,
+			ValidBefore:     r.ValidBefore,
+			Permissions:     r.Permissions,
+			Reserved:        r.Reserved,
+			Signature:       r.Signature,
```

Comment:
A To-Be-Signed certificate won't have a signature.

Reply:
d976e88

Reply:
We will use the x509util.Certificate